plugins/filter_nightfall/nightfall_api.c

```diff
@@ -0,0 +1,476 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2021 The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_filter.h>
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_filter.h>
+#include <fluent-bit/flb_filter_plugin.h>
+#include <fluent-bit/flb_http_client.h>
+#include <fluent-bit/flb_upstream.h>
+#include <fluent-bit/flb_sds.h>
+#include <fluent-bit/flb_time.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/tls/flb_tls.h>
+
+#include "nightfall_api.h"
+
+static int extract_array_fields(struct nested_obj *cur, struct mk_list *stack, 
+                                struct mk_list *payload_list, char *should_pop);
+static int extract_map_fields(struct nested_obj *cur, struct mk_list *stack, 
+                              struct mk_list *payload_list, char *should_pop);
+
+static flb_sds_t build_request_body(struct flb_filter_nightfall *ctx, 
+                             msgpack_object *data)
+{
+    int ret;
+    struct mk_list stack;
+    struct nested_obj *cur;
+    struct nested_obj *new_obj;
+
+    struct mk_list payload_list;
+    struct mk_list *head;
+    struct mk_list *tmp;
+    struct payload *pl;
+    
+    msgpack_sbuffer req_sbuf;
+    msgpack_packer req_pk;
+    flb_sds_t num_str;
+    int num_str_len;
+    flb_sds_t request_body;
+
+    char should_pop = FLB_TRUE;
+
+    new_obj = flb_malloc(sizeof(struct nested_obj));
+    if (!new_obj) {
+        flb_errno();
+        return NULL;
+    }
+    new_obj->obj = data;
+    new_obj->cur_index = 0;
+    new_obj->start_at_val = FLB_FALSE;
+    mk_list_init(&stack);
+    mk_list_add(&new_obj->_head, &stack);
+
+    mk_list_init(&payload_list);
+
+    /* 
+     * Since logs can contain many levels of nested objects, use stack-based DFS here
+     * to extract scannable fields (positive/negative ints, strings)
```

Comment:
wasn't sure if there was already a built in function to go through every field in a log so went with this, happy to be pointed to one if there is